### PR TITLE
[WIP] Manual entry for tests

### PR DESCRIPTION
**Pull request by class4kayaker**
Add entry suggested by #1140

This is a sparse skeleton manual entry for writing and running tests. I have assumed that the section on tests is best included under the section for extending aspect.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -28,6 +28,7 @@
 /doc/aspect.tag
```
Comment (gassmoeller) [inline .gitignore]: Please put these changes in a separate pull request. They are small but not related to the rest.

File: doc/manual/manual.tex [hunk 2]
```diff
@@ -10036,6 +10036,49 @@ \subsection{Extending the basic solver}
 \cite{BHK07,BK99m}.}
 
 
+\subsection{Testing \aspect}
+\label{sec:testing}
+
+\aspect{} makes use of a large suite of tests to ensure correct behavior.
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: Maybe add a sentence that the tests are run automatically with each change to the github repository, but that it is good practice to add new ones, when you add new features.

File: doc/manual/manual.tex [hunk 3]
```diff
@@ -10036,6 +10036,49 @@ \subsection{Extending the basic solver}
...
+
+\subsubsection{Running tests}
+\label{sec:running_tests}
+
+In order to run the tests, it is necessary to have either Diff or Numdiff to
+compare the results to the known good case. While it is possible to use Diff,
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: Maybe add that diff is installed on most Linux systems already and numdiff is usually available as package, so that it is not a severe limitation. Maybe also mention that we use ctest for the testsuite, which should be available on all systems that compiled ASPECT, because it comes with cmake.

File: doc/manual/manual.tex [hunk 4]
```diff
@@ -10036,6 +10036,49 @@ \subsection{Extending the basic solver}
...
+Numdiff is preferred due to being able to more accurately identify whether a
+variation in numerical output is significant.
+
+To run a set of tests, go to the build directory and execute 
+
+\begin{verbatim}
+    ctest -R
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: In fact you do not need to put the `-R` here, `ctest` is sufficient.

File: doc/manual/manual.tex [hunk 5]
```diff
@@ -10036,6 +10036,49 @@ \subsection{Extending the basic solver}
...
+\end{verbatim}
+
+This will run all of the existing tests. To only part of the test suite, it is
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: Other things that might be mentioned here is that by default only a very small subset of tests is run (one I think), unless the `ASPECT_RUN_ALL_TESTS` variable is set to ON during configuration. Also `-j 4` is a useful option to speed up the tests.

Comment (bangerth) [inline doc/manual/manual.tex]: Or you called `make setup_tests`, which I think is the better way to describe what one needs to do.

File: doc/manual/manual.tex [hunk 6]
```diff
@@ -10036,6 +10036,49 @@ \subsection{Extending the basic solver}
...
+
+\begin{verbatim}
+    ctest -R <regex>
+\end{verbatim}
+
+
+\subsubsection{Writing tests}
+\label{sec:writing_tests}
+
+% Copied from email response by gassmoeller, and adjusted
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: No need for the reference :smile: I do not feel like I made a authorship worthy contribution in that mail.

File: doc/manual/manual.tex [hunk 7]
```diff
@@ -10036,6 +10036,49 @@ \subsection{Extending the basic solver}
...
+
+To write a test for a new feature, take one of the existing parameter files in
+the tests/ directory, modify it to use the new feature, check that the new
+feature does what it is supposed to do, and then just add the parameter file to
+the tests directory. You will then need to add another folder to that directory
+that is named like the parameter file, and add the model output files that
+prove that the feature is working (rename log.txt to screen-output for
+historical reasons). The test and output files should be as small and quick to
+run as possible.
```
Comment (gassmoeller) [inline doc/manual/manual.tex]: Here it might be added that the reference test machine might produce slightly different results due to differing libraries, operating systems, and so on. If that happens the reference tester will produce a .diff file on the page of the Github pull request that can be used to modify the checked in test results and resubmit the test results.

**Comment by gassmoeller** [comment]
Thanks for setting this up :+1: whenever you have some time, maybe you could add a few sentences about the things I mentioned.

**Comment by gassmoeller** [comment]
It looks like you already made all the requested changes. For future reference: Write a short comment that says: 'Addressed all comments' or something similar, otherwise we might not notice that this PR is actually ready to be merged. Thanks for contributing! More documentation is always needed! :+1:
